extmod/libmetal/libmetal.mk

```diff
@@ -0,0 +1,48 @@
+# Makefile directives for libmetal
+
+# libmetal is intended to run through a pre-processor (as part of its CMake-based build system).
+# This replicates the basic functionality of the pre-processor, including adding a "micropython"
+# platform that is almost identical to the built-in "generic" platform with a few small changes
+# provided by the files in extmod/libmetal.
+$(BUILD)/openamp: $(BUILD)
+	$(MKDIR) -p $@
+
+$(BUILD)/openamp/metal: $(BUILD)/openamp
+	$(MKDIR) -p $@
+
+$(BUILD)/openamp/metal/config.h: $(BUILD)/openamp/metal $(TOP)/$(LIBMETAL_DIR)/lib/config.h
+	@$(ECHO) "GEN $@"
+	@for file in $(TOP)/$(LIBMETAL_DIR)/lib/*.c $(TOP)/$(LIBMETAL_DIR)/lib/*.h; do $(SED) -e "s/@PROJECT_SYSTEM@/micropython/g" -e "s/@PROJECT_PROCESSOR@/arm/g" $$file > $(BUILD)/openamp/metal/$$(basename $$file); done
+	$(MKDIR) -p $(BUILD)/openamp/metal/processor/arm
+	@$(CP) $(TOP)/$(LIBMETAL_DIR)/lib/processor/arm/*.h $(BUILD)/openamp/metal/processor/arm
+	$(MKDIR) -p $(BUILD)/openamp/metal/compiler/gcc
+	@$(CP) $(TOP)/$(LIBMETAL_DIR)/lib/compiler/gcc/*.h $(BUILD)/openamp/metal/compiler/gcc
+	$(MKDIR) -p $(BUILD)/openamp/metal/system/micropython
+	@$(CP) $(TOP)/$(LIBMETAL_DIR)/lib/system/generic/*.c $(TOP)/$(LIBMETAL_DIR)/lib/system/generic/*.h $(BUILD)/openamp/metal/system/micropython
+	@$(CP) $(TOP)/extmod/libmetal/metal/system/micropython/* $(BUILD)/openamp/metal/system/micropython
+	@$(CP) $(TOP)/extmod/libmetal/metal/config.h $(BUILD)/openamp/metal/config.h
+
+# libmetal's source files.
+SRC_LIBMETAL_C := $(addprefix $(BUILD)/openamp/metal/,\
+	device.c \
+	dma.c \
+	init.c \
+	io.c \
+	irq.c \
+	log.c \
+	shmem.c \
+	softirq.c \
+	version.c \
+	device.c \
+	system/micropython/condition.c \
+	system/micropython/device.c \
+	system/micropython/io.c \
+	system/micropython/irq.c \
+	system/micropython/shmem.c \
+	system/micropython/time.c \
+	)
+
+# These files are generated by the rule above (along with config.h), so we need to make the .c
+# files depend on that step -- can't use the .o files as the target because the .c files don't
+# exist yet.
+$(SRC_LIBMETAL_C): $(BUILD)/openamp/metal/config.h
```

Comment:
This doesn't work as expected, because the qstr preprocessing can run before these source files are generated, and then the build fails (because qstr preprocessing includes these config files).

In particular this PR does not currently build from a fresh start.

The following change fixes it for me:
```
QSTR_GLOBAL_REQUIREMENTS += $(BUILD)/openamp/metal/config.h
```
That will also require updating the comment just above.

Reply:
This builds fine locally, and in the CI. I'm not sure why it's not working for you, but I'll make that change anyway.

Reply:
Is there a build in the CI that builds openamp/libmetal?  I couldn't find one.

Reply:
You're right I thought GIGA was in the CI, should I added ? It would be a good test since it covers many features. Note it still builds here locally.

Reply:
Yes, it would be good to add one of the stm32 Arduino boards to the CI.  You can add that in this PR as a separate commit.

(Maybe you want to add the CI first, before changing this qstr line, to see if the qstr problem is there on the CI.)

Reply:
Yes, and if it doesn't build I'll add the line. Should it be in `ci_stm32_pyb_build` or `ci_stm32_nucleo_build` Or a new bash function like `ci_stm32_misc_build` ?

Reply:
I think we need a new one, because the stm32 build is already taking a while.

Reply:
Hmm, OK, let me try to work out why it doesn't work for me.

Reply:
The issue arises when you build without `-j`, ie just a plain `make`.  I tested on the CI and indeed it also fails when built with just `make`, see https://github.com/micropython/micropython/actions/runs/8292188756/job/22693081383?pr=12961

Reply:
Yes it fails locally too. Adding that line above, just at the end of `libmetal.mk` seems to fix.